### PR TITLE
CLI - functions, hardhat tasks

### PR DESCRIPTION
**Pull request by kanthdefi**
- added periphery dependency
- hardhat config network name changes to use specific network names as `arbgoerli` for testnet and `arb` for mainnet
   - reason: as we are gonna deploy the contracts in polygon and other networks we gotta add support for various networks
- add a Router.sol contract which solely imports the `IRouter` from `periphery` dependency
- update the `LighterConfig` where `OrderBookConfig` was added to contain `Address` and `Id` 
- `CreateOrderEvent` and `SwapExactAmounEvent` classes added to script
- CLI tasks for order creation (all order types excluding performance limit orders) and swapExactSingle (exact-input and exact-output)
- CLI style function and hardhat task to get all `LimitOrders`  and another function to filter then by account (owner of order)

*new fixes and updates:*
* generic function for parseAmount, parseBaseAmount, parseBasePrice added
    * these are when we do CLI based createOrder, swap tasks & user provides amount, price
* these functions check the orderBook metadata and do parsing based on sizeTick, priceTick and token-precisions

* remove redundancy in `createOrder` task which is driven by `orderType`  (optional and defaulted to limitOrder)
    * valid values: 0,2,3 - LimitOrder, FoK, IoC (we don`t have PerformanceLimitOrder at moment)
    * can add if needed

## CLI Outputs

1. SwapExactAmount

```
➜  npx hardhat swapExactInputSingle --orderbookname WETH-USDC --isask false --exactinput 2000 --minoutput 1 --recipient '0xf5306fc60C48E3E2fBf9262D699Cb05C4910e6D9' --unwrap false --network arbgoerli

swapExactInputSingle Transaction: 0x927da390cf32e79f857672aac9b6b3f5f4bc1e09de66ab08ec0a5b5d251bcb10 is successful on OrderBook: WETH-USDC and SwapExactAmount: [
  {
    "sender": "0x42E138965BaCDdF6E1fb2774C55ab84b471c0ab7",
    "recipient": "0xf5306fc60C48E3E2fBf9262D699Cb05C4910e6D9",
    "isExactInput": "true",
    "isAsk": "false",
    "swapAmount0": {
      "type": "BigNumber",
      "hex": "0x0de8d971d3ed2000"
    },
    "swapAmount1": {
      "type": "BigNumber",
      "hex": "0x7735767e"
    }
  }
] emitted
```

2. createLimitOrder

```
➜  npx hardhat createOrder --orderbookname WETH-USDC --ordertype 0 --amount 1 --price 2000 --isask true --network arbgoerli

Create-LimitOrder Transaction: 0x054a5aab89c55d7f02eddf99ce352e8c310d06b73ed8ec43fda7e86739c04cc3 is successful and OrderEvent: {
  "owner": "0xd057E08695d1843FC21F27bBd0Af5D4B06203F48",
  "id": {
    "type": "BigNumber",
    "hex": "0x30"
  },
  "amount0Base": {
    "type": "BigNumber",
    "hex": "0x0186a0"
  },
  "priceBase": {
    "type": "BigNumber",
    "hex": "0x030d40"
  },
  "isAsk": true,
  "orderType": 0
} emitted
```

```
npx hardhat getAllLimitOrders --orderbookname WBTC-USDC --limit 10 --network arbgoerli

orderData queried:

{
   "limit":10,
   "orderCount":5,
   "askOrderCount":3,
   "bidOrderCount":2,
   "askOrders":[
      {
         "id":{
            "type":"BigNumber",
            "hex":"0x1a"
         },
         "isAsk":true,
         "owner":"0xd057E08695d1843FC21F27bBd0Af5D4B06203F48",
         "amount0":{
            "type":"BigNumber",
            "hex":"0x7918"
         },
         "price":{
            "type":"BigNumber",
            "hex":"0x05d21dba00"
         },
         "orderType":0
      },
      {
         "id":{
            "type":"BigNumber",
            "hex":"0x29"
         },
         "isAsk":true,
         "owner":"0xd057E08695d1843FC21F27bBd0Af5D4B06203F48",
         "amount0":{
            "type":"BigNumber",
            "hex":"0x05f5e100"
         },
         "price":{
            "type":"BigNumber",
            "hex":"0x05d21dba00"
         },
         "orderType":0
      },
      {
         "id":{
            "type":"BigNumber",
            "hex":"0x2a"
         },
         "isAsk":true,
         "owner":"0xd057E08695d1843FC21F27bBd0Af5D4B06203F48",
         "amount0":{
            "type":"BigNumber",
            "hex":"0x05f5e100"
         },
         "price":{
            "type":"BigNumber",
            "hex":"0x05d21dba00"
         },
         "orderType":0
      }
   ],
   "bidOrders":[
      {
         "id":{
            "type":"BigNumber",
            "hex":"0x0c"
         },
         "isAsk":false,
         "owner":"0xd057E08695d1843FC21F27bBd0Af5D4B06203F48",
         "amount0":{
            "type":"BigNumber",
            "hex":"0x8dcb38"
         },
         "price":{
            "type":"BigNumber",
            "hex":"0x05b1deb8c0"
         },
         "orderType":0
      },
      {
         "id":{
            "type":"BigNumber",
            "hex":"0x0b"
         },
         "isAsk":false,
         "owner":"0xd057E08695d1843FC21F27bBd0Af5D4B06203F48",
         "amount0":{
            "type":"BigNumber",
            "hex":"0x989680"
         },
         "price":{
            "type":"BigNumber",
            "hex":"0x05b1556480"
         },
         "orderType":0
      }
   ]
}
```

### CODE REVIEW

File: contracts/interface/Router.sol
Comment (alexvelea) [inline contracts/interface/Router.sol:1]: Could we rename this to `IRouter.sol`?

File: config/lighter-config.ts [hunk 1]
```diff
@@ -7,8 +8,14 @@ const lighterConfigs: {
     Router: '0x42E138965BaCDdF6E1fb2774C55ab84b471c0ab7',
     Factory: '0xB15457ea816677564145AF39D92A9aE68fde1E96',
     OrderBooks: {
-      [OrderBookKey.WETH_USDC]: '0x5Fd98c554B29E0DE9DcF09eEd3339367C62b2606',
-      [OrderBookKey.WBTC_USDC]: '0xa791f2eC05b6A6771397B4AE4fAb42E84bb22008',
+      [OrderBookKey.WETH_USDC]: {
+        Address: '0x5Fd98c554B29E0DE9DcF09eEd3339367C62b2606',
+        Id: BigNumber.from(0),
```
Comment (alexvelea) [inline config/lighter-config.ts:13]: Nice!

File: tasks/lighter-orders.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+//priceBase = 197555
+
+// npx hardhat createFillOrKillOrder --orderBookName WETH-USDC --amount 200000 --price 197555 --isAsk true  --network arbgoerli
+task('createFillOrKillOrder')
+  .addParam('orderBookName')
+  .addParam('amount')
+  .addParam('price')
+  .addParam('isask', 'whatever or not order is ask', null, boolean)
+  .setDescription('create FillOrKill Order via Router')
+  .setAction(async ({orderBookName, amount, price, isAsk}, hre) => {
```
Comment (alexvelea) [inline tasks/lighter-orders.ts:24]: I had a bug with this. The variable should be called `isask` lowercase.
Else, it wont be picked up by the hardhat tasks parsing

File: tasks/lighter-orders.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import {IRouter} from '../typechain-types'
+
+
+// WETH-USDC
+// 2 WETH to amount0Base = 20 * 10**18 / 10**14 = 20 * 10**4 = 200000
+//amount0Base = 200000
+// 1975.55 USDC price to priceBase = 1975.55 * 10**6 / 10**4 = 1975.55 * 10**2 = 197555
+//priceBase = 197555
+
+// npx hardhat createFillOrKillOrder --orderBookName WETH-USDC --amount 200000 --price 197555 --isAsk true  --network arbgoerli
```
Comment (alexvelea) [inline tasks/lighter-orders.ts:17]: The user should not be aware of the underlying tick logic
They should use the CLI normaly, like `--amount 0.2 --price 1975.55` and we should do the conversion afterwards

File: tasks/lighter-orders.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  .addParam('isask', 'whatever or not order is ask', null, boolean)
+  .setDescription('create FillOrKill Order via Router')
+  .setAction(async ({orderBookName, amount, price, isAsk}, hre) => {
+    const lighterConfig = await getLighterConfig()
+    const routerContract = getRouterAt(lighterConfig.Router, hre)
+    const orderBookId = lighterConfig.OrderBooks[orderBookName as OrderBookKey]?.Id as BigNumber
+    const amountBase = BigNumber.from(amount)
+    const priceBase = BigNumber.from(price)
+    const tx = await (await routerContract).createFoKOrder(orderBookId, amountBase, priceBase, isAsk)
+    await tx.wait()
```
Comment (alexvelea) [inline tasks/lighter-orders.ts:31]: Looking good.
Maybe print a message with the response?
I think I think it's hard to get the exact swapped amounts from the TX, but maybe it's possible?
Would be nice to have.
Else, we can just print something like (succesfully swapped ...)

Comment (kanthdefi) [reply tasks/lighter-orders.ts:27]: updated

File: tasks/lighter-orders.ts (hunk 3 above)
Comment (kanthdefi) [reply tasks/lighter-orders.ts]: updated @alexvelea

File: shared/event-util.ts [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    public owner: string,
+    public id: BigNumber,
+    public amount0Base: BigNumber,
+    public priceBase: BigNumber,
+    public isAsk: boolean,
+    public orderType: OrderType
+  ) {}
+}
+
+export const getCreateOrderEvent = async (
```
Comment (alexvelea) [inline shared/event-util.ts:17]: - I think this should return a list of `CreateOrderEvent` as multiple orders can be created in one TX
- I'm not sure provider is required, as I think you can get it from `hre`

File: shared/event-util.ts [hunk 6]
```diff
@@ -0,0 +1,166 @@
+import '@nomiclabs/hardhat-ethers'
+import {Provider} from '@ethersproject/providers'
+import {OrderType, getOrderBookAt, getOrderTypeFromValue, getTransactionReceipt} from '../shared'
+import {BigNumber, ethers} from 'ethers'
+
+export class CreateOrderEvent {
```
Comment (alexvelea) [inline shared/event-util.ts:6]: I think it would be easier to have this as interfaces instead of classes, since we don't add any methods to them

Comment (kanthdefi) [reply shared/event-util.ts]: updated to interface

File: shared/event-util.ts (hunk 5 above)
Comment (kanthdefi) [reply shared/event-util.ts:15]: updated

**Review by alexvelea — APPROVED** [review]

**Review by muratekici — APPROVED** [review]
lgtm overall

- should we fetch all the pages instead of just one page?
- should we throw an error if given precision is incorrect?

@alexvelea @kanthdefi

**Comment by alexvelea** [comment]
> * should we fetch all the pages instead of just one page?

We went with a big limit
Should be enough, until we deploy a ViewContract or similar

> * should we throw an error if given precision is incorrect?

I'll add them

**Pull request merged**